File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.py
# Q2007m, runtime 84.56%, py
class Solution:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        ans = []
        if len(changed) % 2 == 1: return []
        changed.sort()
        odd2grp = defaultdict(dict)
        for n in changed:
            odd = n
            while odd != 0 and odd % 2 == 0:
                odd //= 2
            odd2grp[odd][n] = odd2grp[odd].get(n, 0) + 1
        for odd, grp in odd2grp.items():
            vals = list(grp.keys())
            vals.sort()
            ind = 0
            while ind < len(vals):
                p = vals[ind]
                if grp[p] <= 0: # skip cnt==0
                    ind += 1
                    continue
                if 2 * p not in grp:
                    return []
                if p == 0 and grp[2 * p] <= 1 or p >= 0 and grp[2 * p] <= 0:
                    return []
                grp[p] -= 1
                grp[2 * p] -= 1
                ans.append(p)
                if grp[p] <= 0:
                    ind += 1
        return ans
        

File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.ts
// Q2007m, runtime 88.89%, TS
function findOriginalArray(changed: number[]): number[] {
    let ans: number[] = []; // soll init
    let cnt = new Map();
    // js sort() is really problematic
    changed.sort((a, b) => a - b);
    for (let n of changed) { // gramma here is of
        if (cnt.has(n))
            cnt.set(n, cnt.get(n) + 1);
        else
            cnt.set(n, 1);
    }
    for (let n of changed) {
        if (cnt.get(n) == 0)
            continue;
        if (!cnt.has(n * 2))
            return [] as number[];
        if (n == 0 && cnt.get(n) <= 1 || n >= 1 && cnt.get(n) <= 0)
            return [] as number[];
        cnt.set(n, cnt.get(n) - 1);
        cnt.set(n * 2, cnt.get(n * 2) - 1);
        ans.push(n);
    }
    return ans;
};